waitable_atomic/waitable_atomic.h

```diff
@@ -0,0 +1,324 @@
+    friend class TemplatedScopedAccessor<const BasicImpl>;
+
+    inline ImmutableAccessor ImmutableScopedAccessor() const { return ImmutableAccessor(this); }
+
+    inline MutableAccessor MutableScopedAccessor() { return MutableAccessor(this); }
+
+    inline void Notify() { data_condition_variable_.notify_all(); }
+
+    inline void UseAsLock(std::function<void()> f) const {
+      std::unique_lock<std::mutex> lock(T_DATA::data_mutex_);
+      f();
+    }
+
+    inline bool Wait(std::function<bool(const T_DATA&)> predicate) const {
+      std::unique_lock<std::mutex> lock(data_mutex_);
+      if (!predicate(data_)) {
+        const T_DATA& data = std::ref(data_);
+        data_condition_variable_.wait(lock, [&predicate, &data] { return predicate(data); });
+      }
+      return true;
+    }
+
+    inline void ImmutableUse(std::function<void(const T_DATA&)> f) const {
+      auto scope = ImmutableScopedAccessor();
+      f(*scope);
+    }
+
+    inline void MutableUse(std::function<void(T_DATA&)> f) {
+      auto scope = MutableScopedAccessor();
+      f(*scope);
+    }
+
+    inline void PotentiallyMutableUse(std::function<bool(T_DATA&)> f) {
+      auto scope = MutableScopedAccessor();
+      if (!f(*scope)) {
+        scope.MarkAsUnmodified();
+      }
+    }
+
+    inline T_DATA GetValue() const { return *ImmutableScopedAccessor(); }
+
+    inline void SetValue(const T_DATA& data) { *MutableScopedAccessor() = data; }
+
+    inline void SetValueIf(std::function<bool(const T_DATA&)> predicate, const T_DATA& data) {
+      auto a = MutableScopedAccessor();
+      if (predicate(*a)) {
+        *a = data;
+      } else {
+        a.MarkAsUnmodified();
+      }
+    }
+
+   protected:
+    T_DATA data_;
+    mutable std::mutex data_mutex_;
+    mutable std::condition_variable data_condition_variable_;
+
+   private:
+    BasicImpl(const BasicImpl&) = delete;
+    void operator=(const BasicImpl&) = delete;
+    BasicImpl(BasicImpl&&) = delete;
+  };
+
+  class IntrusiveImpl : public BasicImpl, public IntrusiveClient::Interface {
+   public:
+    typedef DATA T_DATA;
+    enum { IS_INTRUSIVE = true };
+
+    explicit IntrusiveImpl(CustomWaitableAtomicDestructor* destructor_ptr = nullptr)
+        : destructor_ptr_(destructor_ptr) {
+      RefCounterTryIncrease();
+    }
+
+    inline explicit IntrusiveImpl(const T_DATA& data, CustomWaitableAtomicDestructor* destructor_ptr = nullptr)
+        : BasicImpl(data), destructor_ptr_(destructor_ptr) {
+      RefCounterTryIncrease();
+    }
+
+    inline virtual ~IntrusiveImpl() {
+      {
+        std::lock_guard<std::mutex> guard(BasicImpl::data_mutex_);
+        not_destructing_ = false;
+        BasicImpl::Notify();
+      }
+      RefCounterDecrease();
+      if (destructor_ptr_) {
+        destructor_ptr_->WaitableAtomicDestructing();
+      }
+      ref_count_is_nonzero_mutex_.lock();
+    }
+
+    inline bool Wait(std::function<bool(const T_DATA&)> predicate) const {
```

Comment:
Locking the main `data_mutex_` and checking for `!destructing_` as the first condition of the predicate should take care of that, shoudn't they?

Could you describe the case in more detail, please?